cuML/src/metrics/trustworthiness.cu

```diff
@@ -0,0 +1,202 @@
+
+            cudaStream_t stream = h.getStream();
+            auto d_alloc = h.getDeviceAllocator();
+
+            size_t workspaceSize = 0; // EucUnexpandedL2Sqrt does not reauire workspace (may need change for other distances)
+            typedef cutlass::Shape<8, 128, 128> OutputTile_t;
+            bool bAllocWorkspace = false;
+
+            math_t* d_pdist_tmp = (math_t*)d_alloc->allocate(TMP_SIZE * sizeof(math_t), stream);
+            int* d_ind_X_tmp = (int*)d_alloc->allocate(TMP_SIZE * sizeof(int), stream);
+
+            long* ind_X_embedded = get_knn_indexes(h, X_embedded,
+                n, d, n_neighbors + 1);
+
+            double t_tmp = 0.0;
+            double t = 0.0;
+            double* d_t = (double*)d_alloc->allocate(sizeof(double), stream);
+
+            int toDo = n;
+            while (toDo > 0)
+            {
+                int batchSize = min(toDo, MAX_BATCH_SIZE);
+                // Takes at most MAX_BATCH_SIZE vectors at a time
+
+                distance<distance_type, math_t, math_t, math_t, OutputTile_t>
+                        (&X[(n - toDo) * m], X,
+                        d_pdist_tmp,
+                        batchSize, n, m,
+                        (void*)nullptr, workspaceSize,
+                        stream
+                );
+                CUDA_CHECK(cudaPeekAtLastError());
+
+                sortColumnsPerRow(d_pdist_tmp, d_ind_X_tmp,
+                                    batchSize, n,
+                                    bAllocWorkspace, NULL, workspaceSize,
+                                    stream);
+                CUDA_CHECK(cudaPeekAtLastError());
+
+                t_tmp = 0.0;
+                updateDevice(d_t, &t_tmp, 1, stream);
+
+                int work = batchSize * n_neighbors;
+                int n_blocks = work / N_THREADS + 1;
+                compute_rank<<<n_blocks, N_THREADS, 0, stream>>>(d_ind_X_tmp,
+                        &ind_X_embedded[(n - toDo) * (n_neighbors+1)],
+                        n,
+                        n_neighbors,
+                        batchSize * n_neighbors,
+                        d_t);
+                CUDA_CHECK(cudaPeekAtLastError());
+
+                updateHost(&t_tmp, d_t, 1, stream);
+                t += t_tmp;
+
+                toDo -= batchSize;
+            }
+
+            t = 1.0 - ((2.0 / ((n * n_neighbors) * ((2.0 * n) - (3.0 * n_neighbors) - 1.0))) * t);
+
+            d_alloc->deallocate(ind_X_embedded, n * (n_neighbors + 1) * sizeof(long), stream);
+            d_alloc->deallocate(d_pdist_tmp, TMP_SIZE * sizeof(math_t), stream);
+            d_alloc->deallocate(d_ind_X_tmp, TMP_SIZE * sizeof(int), stream);
+            d_alloc->deallocate(d_t, sizeof(double), stream);
+
+            return t;
+        }
+
+        template<typename math_t>
+        double trustworthiness_score(const cumlHandle& h, math_t* X,
+            math_t* X_embedded, int n, int m, int d,
+            int n_neighbors, int metric)
```

Comment:
Do we need this function at all? Could this be accomplished by using the following for the template in the previous function?

```
template<typename math_t = float, DistanceType distance_type = EucUnexpandedL2Sqrt>
``` 

and having a simple validation for the distance type inside the above function?

Reply:
Cython is not normally able to use values (as opposed to types) as template parameters. I just managed to use a Cython trick to do it. Unfortunately, I had to make a copy of the DistanceType enum because if I included the original file directly I got Cython compile errors because of Cuda stuff in there. The problem is that if the enum needs modifications, the trustworthiness_c.h file must be updated too.

Reply:
I won't hold up the merge of this PR with more changes, however, I believe this problem would be solved if we moved the trustworthiness score into the primitives layer with the full templated API and only use the functions in the algorithms layer (the `cuML` portion of the codebase) to expose the primitives to cython.

Reply:
Ok, I'll do that on monday.

Reply:
I am in a little bit of a dilemma here. My function uses a cumlHandle as first parameter. Should I include things from ML namespace like cumlHandle into primitives code ?

Reply:
I don't think this will be too hard of a change. Here's a good example taken from our TSVD implementation, which passes `handle.getDeviceAllocator()` into the primitives layer. You would pass `handle.getStream()` as well. 

https://github.com/rapidsai/cuml/blob/branch-0.8/cuML/src/tsvd/tsvd.h#L222
https://github.com/rapidsai/cuml/blob/branch-0.8/cuML/src/tsvd/tsvd.h#L112

Reply:
Ok thx, but I have a second problem. I need to access the KNN which is also part of ML namespace and has its header file in cuML.

Reply:
We will be moving the kNN down to our `selection` primitives in the near future. This depends on a few other moving pieces, some of which are currently in flight. 

A PR will be getting merged soon which should better consolidate the `thirdparty` dependencies across the primitives and the algorithms. This should bring us a little closer to making the build changes necessary for the switch. 

Until then, you've done a great job on this PR and I see a couple of options to enable us to proceed-
1. We can open a new Github issue for moving this to the primitives layer (perhaps in 0.9 timeframe) and merge this PR as-is. 
2. We can have the primitive accept the knn_dist/knn_ind arrays directly for now, which will allow the kNN to remain in the algorithms layer until we are able to move it. 

Do either one of these sound reasonable to you?

Reply:
Thanks, I think it would be cleaner and simpler to just merge the PR as-is and to wait for the other PR to be merged before making the change. Yes, we can open a Github issue for that.

Reply:
Great! Once these conflicts are fixed, I'll merge this in. Thanks again, Victor!